app/lib/differential_expression_service.rb

```diff
@@ -1,5 +1,116 @@
+  #   - +user+ (User) => Corresponding user, will default to study owner
+  #
+  # * *yields*
+  #   - (IngestJob) => Differential expression job in PAPI for each valid cluster/annotation combination
+  #
+  # * *returns*
+  #   - (Integer) => Number of DE jobs yielded
+  #
+  # * *raises*
+  #   - (ArgumentError) => if requested study cannot run any DE jobs
+  def self.run_differential_expression_on_all(study_accession, user: nil)
+    study = Study.find_by(accession: study_accession)
+    validate_study(study)
+    available_annotations = []
+
+    metadata = study.cell_metadata.where(annotation_type: 'group', is_differential_expression_enabled: false)
+                    .select(&:can_visualize?)
+    available_annotations += metadata.map do |meta|
+      { annotation_name: meta.name, annotation_type: meta.annotation_type, annotation_scope: 'study' }
+    end
+
+    cell_annotations = []
+    study.cluster_groups.each do |cluster|
+      next if cluster.cell_annotations.empty?
+
+      cell_annots = cluster.cell_annotations.select do |annot|
+        annot['type'] == 'group' &&
+          cluster.can_visualize_cell_annotation?(annot) &&
+          !annot[:is_differential_expression_enabled]
+      end
+
+      cell_annots.each do |annot|
+        annot[:cluster_file_id] = cluster.study_file.id # for checking associations later
+      end
+
+      cell_annotations += cell_annots
+    end
+
+    available_annotations += cell_annotations.map do |annot|
+      {
+        annotation_name: annot[:name],
+        annotation_type: annot[:type],
+        annotation_scope: 'cluster',
+        cluster_file_id: annot[:cluster_file_id]
+      }
+    end
+    raise ArgumentError, "#{study_accession} does not have any valid annotations" if available_annotations.empty?
+
+    log_message "#{study_accession} has data available for DE; validating inputs"
+    requested_user = user || study.user
+
+    job_count = 0
+    study.cluster_ordinations_files.each do |cluster_file|
+      available_annotations.each do |annotation|
+        begin
+          # skip if this is a cluster-based annotation and is not available on this cluster file
+          next if annotation[:scope] == 'cluster' && annotation[:cluster_file_id] != cluster_file.id
+
+          annotation_params = annotation.deep_dup # make a copy so we don't lose the association next time we check
+          annotation_params.delete(:cluster_file_id)
+          job_identifier = "#{study_accession}: #{cluster_file.name} (#{annotation_params.values.join('--')})"
+          log_message "Checking DE job for #{job_identifier}"
+          DifferentialExpressionService.run_differential_expression_job(
+            cluster_file, study, requested_user, **annotation_params
+          )
+          log_message "DE job for #{job_identifier} successfully launched"
+          job_count += 1
+        rescue ArgumentError => e
+          log_message "Skipping DE job for #{job_identifier} due to: #{e.message}"
```

Comment:
Is this error worth logging to Sentry?

Reply:
No, because it's not actually an "error" that we care about - it's a validation error as opposed to something failing.  The way the code is written, it either submits a job or emits and `ArgumentError` because the requested parameters do not meet the requirements for a valid DE job.  I wrote it this way so that it was more explicit when a job can't run, rather than exiting silently.